crates/oxc_query/src/properties.rs

```diff
@@ -0,0 +1,584 @@
+use std::convert::Into;
+
+use oxc_ast::ast::{BindingPatternKind, Expression, MemberExpression};
+use oxc_span::GetSpan;
+use trustfall::{
+    provider::{resolve_property_with, ContextIterator, ContextOutcomeIterator, ResolveInfo},
+    FieldValue,
+};
+
+use super::vertex::Vertex;
+use crate::{
+    util::{
+        accessibility_to_string, jsx_attribute_name_to_string, jsx_attribute_to_constant_string,
+        jsx_element_name_to_string,
+    },
+    vertex::InterfaceExtendVertex,
+    Adapter,
+};
+
+fn interface_extend_implem<'a, 'b: 'a>(
+    contexts: ContextIterator<'a, Vertex<'b>>,
+) -> ContextOutcomeIterator<'a, Vertex<'b>, FieldValue> {
+    resolve_property_with(contexts, |v| {
+        match v
+            .as_interface_extend()
+            .unwrap_or_else(|| {
+                panic!("expected to have an interfaceextend vertex, instead have: {v:#?}")
+            })
+            .as_ref()
+        {
+            InterfaceExtendVertex::Identifier(ident) => ident.name.to_string(),
+            InterfaceExtendVertex::MemberExpression(first_membexpr) => {
+                let MemberExpression::StaticMemberExpression(static_membexpr) = first_membexpr
+                else {
+                    unreachable!("TS:2499")
+                };
+                let mut parts = vec![static_membexpr.property.name.to_string()];
+                let mut membexpr = first_membexpr.object();
+                while let Expression::MemberExpression(expr) = membexpr {
+                    let MemberExpression::StaticMemberExpression(static_membexpr) = &expr.0 else {
+                        unreachable!("TS:2499")
+                    };
+                    parts.push(static_membexpr.property.name.to_string());
+                    membexpr = expr.object();
+                }
+
+                let Expression::Identifier(ident) = membexpr else { unreachable!("TS:2499") };
+                parts.push(ident.name.to_string());
+
+                parts.reverse();
+
+                parts.join(".")
+            }
+        }
+        .into()
+    })
+}
+
+pub(super) fn resolve_assignment_type_property<'a, 'b: 'a>(
+    contexts: ContextIterator<'a, Vertex<'b>>,
+    property_name: &str,
+    _resolve_info: &ResolveInfo,
+) -> ContextOutcomeIterator<'a, Vertex<'b>, FieldValue> {
+    match property_name {
+        "assignment_to_variable_name" => resolve_property_with(contexts, |v| {
+            let Vertex::AssignmentType(BindingPatternKind::BindingIdentifier(ident)) = v else {
```

Comment:
I would switch this to an as, but then the BindingPatternKind will still need to be pattern-matched out of the vertex so I guess it's fine to leave it like this